File: oolab/src/main/java/agh/ics/oop/model/Vector2d.java
package agh.ics.oop.model;

import java.util.Objects;

import static java.lang.Integer.max;
import static java.lang.Integer.min;

public class Vector2d {
    private final int x, y;

    public Vector2d(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }
    public int getY() {
        return y;
    }

    public boolean precedes(Vector2d other) {
        return x <= other.getX() && y <= other.getY();
    }

    public boolean follows(Vector2d other) {
        return x >= other.getX() && y >= other.getY();
    }

    public Vector2d add(Vector2d other) {
        return new Vector2d(x+other.getX(), y+other.getY());
    }

    public Vector2d subtract(Vector2d other) {
        return new Vector2d(x-other.getX(), y-other.getY());
    }

    public Vector2d upperRight(Vector2d other) {
        return new Vector2d(max(x, other.getX()), max(y, other.getY()));
    }

    public Vector2d lowerLeft(Vector2d other) {
        return new Vector2d(min(x, other.getX()), min(y, other.getY()));
    }

    public Vector2d opposite() {
        return new Vector2d(-x, -y);
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof Vector2d)) return false;

        Vector2d that = (Vector2d) other;
        return (x == that.x && y == that.y);
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public String toString() {
        return "(%d,%d)".formatted(x, y);
    }
}

File: oolab_kotlin/src/main/kotlin/agh/ics/oop/model/Vector2d.kt
package agh.ics.oop.model

import kotlin.math.max
import kotlin.math.min

data class Vector2d(val x: Int, val y: Int) {
    fun precedes(other: Vector2d): Boolean = x <= other.x && y <= other.y

    fun follows(other: Vector2d): Boolean = x >= other.x && y >= other.y

    fun add(other: Vector2d): Vector2d = Vector2d(x + other.x, y + other.y)

    fun subtract(other: Vector2d): Vector2d = Vector2d(x - other.x, y - other.y)

    fun upperRight(other: Vector2d): Vector2d = Vector2d(max(x, other.x), max(y, other.y))

    fun lowerLeft(other: Vector2d): Vector2d = Vector2d(min(x, other.x), min(y, other.y))

    fun opposite(): Vector2d = Vector2d(-x, -y)

    override fun toString(): String = "($x,$y)"
    operator fun unaryMinus() = opposite()
    operator fun plus(other: Vector2d) = add(other)
    operator fun minus(other: Vector2d) = subtract(other)
}

fun MapDirection.toUnitVector(): Vector2d = when(this) {
    MapDirection.NORTH -> Vector2d(0, 1)
    MapDirection.SOUTH -> Vector2d(0, -1)
    MapDirection.EAST -> Vector2d(1, 0)
    MapDirection.WEST -> Vector2d(-1, 0)
}